bench/many_deep.html

```diff
@@ -0,0 +1,51 @@
+<!DOCTYPE html>
+<html>
+
+<head></head>
+
+<body>
+  <script type="module">
+    import { hydrateShadowRoots as manualWalkHydration } from '../_implementation/manual_walk.js';
+    import { hydrateShadowRoots as querySelectorHydration } from '../_implementation/queryselectorall.js';
+
+    function benchmark(hydrateMethod) {
+      let hydrateShadowRoots;
+      if (hydrateMethod === 'manualWalk') {
+        hydrateShadowRoots = manualWalkHydration;
+      } else if (hydrateMethod === 'querySelector') {
+        hydrateShadowRoots = querySelectorHydration;
+      } else {
+        return 9 * 1000 * 1000;
+      }
+      let html = '';
+      const n = 1000;
+      for (let i = 0; i < n; i++) {
+        html += `<div>
+          <template shadowroot="open">
+        `;
+      }
+      for (let i = 0; i < n; i++) {
+        html += `
+              <div>Hello world</div>
+            </template>
+          </div>
+        `;
+      }
+
+      const start = performance.now();
+      document.body.innerHTML = html;
+      hydrateShadowRoots(document.body);
+      const duration = performance.now() - start;
+      if (document.querySelector('template') === null) {
+        return duration;
+      } else {
+        // hydration failed, still see templates in the document!
+        return 9 * 1000 * 1000;
+      }
+    }
+
+    window.tachometerResult = benchmark(window.location.search.slice(1));
```

Comment:
I'm not sure exactly what would be different with static HTML, but we might ultimately care more about that and using Tach's FCP feature. ie, we could have a generator create a static file with lots of shadow roots instead of doing it client-side.

Reply:
+1, good approach for follow up